libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/target/SyncUpTarget.kt

```diff
@@ -0,0 +1,573 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.salesforce.androidsdk.mobilesync.target
+
+import com.salesforce.androidsdk.mobilesync.manager.SyncManager
+import com.salesforce.androidsdk.mobilesync.manager.SyncManager.MobileSyncException
+import com.salesforce.androidsdk.mobilesync.util.Constants
+import com.salesforce.androidsdk.rest.RestRequest
+import com.salesforce.androidsdk.smartstore.store.SmartStore
+import com.salesforce.androidsdk.util.JSONObjectHelper
+import org.json.JSONArray
+import org.json.JSONException
+import org.json.JSONObject
+import java.io.IOException
+import java.net.HttpURLConnection
+
+/**
+ * Target for sync up:
+ * - what records to upload to server
+ * - how to upload those records
+ *
+ * During a sync up, sync manager does the following:
+ *
+ * 1) it calls getIdsOfRecordsToSyncUp to get the ids of records to sync up
+ *
+ * 2) for each id it does:
+ *
+ * a) if calls getFromLocalStore to get the record itself
+ *
+ * b) if merge mode is leave-if-changed, it calls isNewerThanServer, if that returns false, it goes to the next id
+ *
+ * c) otherwise it does one of the following three operations:
+ * - calls deleteOnServer if isLocallyDeleted returns true for the record (unless it is also locally created, in which case it gets deleted locally right away)
+ * if successful or not found is returned, it calls deleteFromLocalStore to delete record locally
+ *
+ * - calls createOnServer if isLocallyCreated returns true for the record
+ * if successful, it updates the id to be the id returned by the server and then calls cleanAndSaveInSmartstore to reset local flags and save the record locally
+ *
+ * - calls updateOnServer if isLocallyUpdated returns true for the record
+ * if successful, it calls cleanAndSaveInSmartstore to reset local flags and save the record
+ * if not found and merge mode is overwrite, it calls createOnServer to recreate the record on the server
+ *
+ */
+open class SyncUpTarget : SyncTarget {
+    // Fields
+    @JvmField
+    var createFieldlist: List<String>?
+    @JvmField
+    var updateFieldlist: List<String>?
+
+    /**
+     * @return The field name of an external id field of the record.  Default to null.
+     */
+    var externalIdFieldName: String?
+        protected set
+
+    // Last sync error
+    protected var lastError: String? = null
+    /**
+     * Construct SyncUpTarget
+     */
+    /**
+     * Construct SyncUpTarget
+     */
+    /**
```

Comment:
Optional: It looks like there's two extra comment blocks here.

Reply:
Java -> Kotlin conversion combined multiple method into a single method with default parameter values (and @JvmOverloads annotations). But it kept all the java doc comments ;-) Will cleanup.

Reply:
See https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/2459/commits/667e0fb3b229c4e525a5f6b5a96e29b8357d9ecc